Electron provider with msal (#11289)
microsoftgraph/microsoft-graph-docs: concepts/toolkit/get-started/build-an-electron-app.md

description: "Get started using the Microsoft Graph Toolkit in an Electron application."
localization_priority: Normal
author: amrutha95
---

# Use the Microsoft Graph Toolkit with Electron

This article describes the step-by-step process of using the Microsoft Graph Toolkit to create an Electron app and connect it to Microsoft 365. After completing the steps, you'll have a Electron app that shows the upcoming appointments of the currently signed in user from Microsoft 365.

## Create an Electron app 
Create a new Electron app by cloning this git repository ([electron-quick-start-typescript](https://github.com/electron/electron-quick-start-typescript)). This will create a new Electron app using TypeScript, which will help you write more robust code and avoid runtime errors.

```cmd
git clone https://github.com/electron/electron-quick-start-typescript
```

Change the working directory to the newly created app and install all dependencies.

```cmd
cd electron-quick-start-typescript
npm install
```

Install the '@microsoft/mgt-components' package that contains all the Microsoft Graph connected web components.

```cmd
npm i @microsoft/mgt-components
```

> Install the `@microsoft/mgt-electron-provider` and `@microsoft/mgt-element` npm packages as well. These will allow you to provide authentication for your app using MSAL, and further be able to use the MGT components.

Reviewer: ```suggestion
Install the `@microsoft/mgt-electron-provider` and `@microsoft/mgt-element` npm packages as well. These will allow you to provide authentication for your app using MSAL, and further be able to use the Microsoft Graph Toolkit components.
```

```
Next, add these lines of code in the `createWindow()` function to initialize the ElectronAuthenticator, right after where `mainWindow` is declared. Replace `<your_client_id>` with the client id from your app registration.

```ts
const config: MsalElectronConfig = {
  clientId: '<your_client_id>',
  mainWindow: mainWindow, //This is the BrowserWindow instance that requires authentication
  scopes: [
    'user.read',
        'user.read',
        'people.read',
        'user.readbasic.all',
        'contacts.read',
        'presence.read.all',
        'presence.read',
        'user.read.all',
        'calendars.read'
  ],
};
ElectronAuthenticator.initialize(config);
```

### Set nodeIntegration to true
 
In main.ts where the new instance of BrowserWindow is created, make sure that you set `nodeIntegration` to `true` under webPreferences. If you skip this step, you may run into a ```Uncaught ReferenceError: require is not defined``` error. To keep this simple, remove preloading scripts if any.

```ts
const mainWindow = new BrowserWindow({
  height: 600,
  webPreferences: {
    nodeIntegration: true //Set this to true
  },
  width: 800
});
```
 

>  ### Add MGT components to your HTML page

Reviewer: ```suggestion
### Add components to your HTML page
```

 
 #### Add the webpacking script in ```package.json```
 
 Add the following under ```scripts``` in your ```package.json```
 ```json
"scripts": {
   "webpack": "webpack",
   "start": "npm run webpack && electron dist/main.js"
 }
                
 ```
 
 #### Run your app
 
 ```cmd
 npm start
 ```

### [Optional] Add token caching capabilities to your app and enable silent log-ins.

[MSAL Node](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-node) supports an in-memory cache by default and provides the ICachePlugin interface to perform cache serialization, but does not provide a default way of storing the token cache to disk. If you need persistent cache storage to enable silent log-ins or cross-platform caching, we recommend using the default implementation provided by MSAL Node [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/msal-node-extensions). You can import this plugin, and pass the instance of the cache plugin while initializing ElectronAuthenticator.

```ts
let config: MsalElectronConfig = {
  ...

>   cachePlugin: new PersistenceCachePlugin(filePersistence)

Reviewer 2: what is `filePersisience`?

To create the app in Azure Active Directory:

1. Go to the Azure portal at https://portal.azure.com.
1. From the menu, select **Azure Active Directory**.
1. From the Azure Active Directory menu, select **App registrations**.
1. From the top menu, select the **New registration** button.
1. Enter the name for your app; for example, `My Electron-App`.
1. For the type of [supported account types](/azure/active-directory/develop/single-and-multi-tenant-apps#who-can-sign-in-to-your-app), select **Accounts in any organizational directory (Any Azure AD directory - Multitenant) and personal Microsoft accounts (e.g. Skype, Xbox)**.
1. In the **Redirect URI** field, in the dropdown, select **Public client/native (mobile & desktop)**, and in the URL field, enter `msal://redirect`.
1. Confirm changes by selecting the **Register** button.
1. Go to your application registration.
1. Verify that you are on the **Overview** page.
1. From the **Essentials** section, copy the value of the **Application (client) ID** property.

## Configure the Microsoft Graph Toolkit authentication provider

> ### Initializing ElectronProvider in your renderer process (renderer.ts, Front end)
> 
> The ElectronProvider is responsible for communicating with ElectronAuthenticator (in the main process) to request access tokens and receive information regarding logged in state that are required for the mgt components to work. 
> 

Reviewer 2: ```suggestion
### Initializing ElectronProvider in your renderer process

The `ElectronProvider` is responsible for communicating with `ElectronAuthenticator` (in the main process) to request access tokens and receive information regarding logged in state that are required for the mgt components to work. 

To initialize the `ElectronProvider`, add the following code to the *src/rendered.ts* file:
```

```ts
import {Providers} from '@microsoft/mgt-element';
import {ElectronProvider} from '@microsoft/mgt-electron-provider/dist/Provider';

> import '@microsoft/mgt-components';

Reviewer 2: ```suggestion
// import the mgt components so we can use them in our html
import '@microsoft/mgt-components';
```

// initialize the auth provider globally
Providers.globalProvider = new ElectronProvider();
```

> ### Initializing ElectronAuthenticator in your main process (main.ts, Back end)
> 
> The ElectronAuthenticator is responsible for setting up the configuration variables for MSAL authentication, acquiring access tokens and communicating with ElectronProvider.
> Initialize the ElectronAuthenticator in the main process and set up the configuration variables such as client-id.
> First, import `ElectronAuthenticator` and  `MsalElectronConfig` from `@microsoft/mgt-electron-provider`"
> 

Reviewer 2: ```suggestion
### Initializing ElectronAuthenticator in your main process

The `ElectronAuthenticator` is responsible for setting up the configuration variables for MSAL authentication, acquiring access tokens and communicating with the `ElectronProvider`. Let's Initialize the `ElectronAuthenticator` in the main process and set up the configuration variables such as client-id and required scopes. 

First, open *src/main.ts* and import `ElectronAuthenticator` and  `MsalElectronConfig` from `@microsoft/mgt-electron-provider` at the top of the page:

```

```
Next, add these lines of code in the `createWindow()` function to initialize the ElectronAuthenticator, right after where `mainWindow` is declared. 
```ts
const config: MsalElectronConfig = {
  clientId: '<your_client_id>',
  mainWindow: mainWindow, //This is the BrowserWindow instance that requires authentication
  scopes: [
    'user.read',
        'user.read',
        'people.read',
        'user.readbasic.all',
        'contacts.read',
        'presence.read.all',
        'presence.read',
        'user.read.all',
        'calendars.read'
  ],
};
ElectronAuthenticator.initialize(config);
```

### Set nodeIntegration to true
 
In main.ts where the new instance of BrowserWindow is created, make sure that you set `nodeIntegration` to `true` under webPreferences. If you skip this step, you may run into a ```Uncaught ReferenceError: require is not defined``` error. To keep this simple, remove preloading scripts if any.

```ts
const mainWindow = new BrowserWindow({
  height: 600,
  webPreferences: {
    nodeIntegration: true //Set this to true
  },
  width: 800
});
```
 
 ### Add MGT components to your HTML page
 

>  Here is an example index.html:

Reviewer 2: ```suggestion
Let's add some content to our app. We can now use the Microsoft Graph toolkit components in our *index.html* page and show the user's agenda. Replace the content of the *index.html* page with the following:
```

 
 ```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <title>Sample Electron-MGT App</title>
  </head>
  <body>
    <mgt-login></mgt-login>
    <mgt-agenda group-by-day></mgt-agenda>
    <script type="module" src="./dist/renderer.js"></script>
  </body>
</html>
 ```
 
 
 ### Bundle your app using webpack
 

>  We recommend this step to avoid running into errors by bundling all the modular dependencies in this project.

Reviewer 2: ```suggestion
 Before we can run the app, we will need to bundle our code to ensure all of our modular dependencies are included in the final payload. If you are already bundling your app code, you can skip this step.
```

 
 #### Install webpack
 
 ```cmd 
 npm install webpack webpack-cli ts-loader --save-dev
 ```
 

>  Create a webpack.config.js file in the root folder of your project.
>  
>  #### webpack.config.js

Reviewer 2: ```suggestion
 #### webpack.config.js
 
Create a new *webpack.config.js* file in the root folder of your project, and paste the following configuration:
 
```
